core/blockchain.go

```diff
@@ -153,6 +153,17 @@ var defaultCacheConfig = &CacheConfig{
 	SnapshotWait:   true,
 }
 
+// BlockchainLogger is used to collect traces during chain processing.
+// Please make a copy of the referenced types if you intend to retain them.
+type BlockchainLogger interface {
+	vm.EVMLogger
+	state.StateLogger
+	OnBlockStart(*types.Block)
+	OnBlockEnd(td *big.Int, err error)
```

Comment:
Block finalization info is something of great importance here when it comes to dealing with chain moving forward. We use finality of blocks to implement pruning of now forked data as well as handling some persistence system.

Is there any problem in doing `bc.CurrentFinalBlock` at call site when `OnBlockEnd` is invoked?

Reply:
Ok! we can do that. Or alternatively have a `OnChainUpdate(safeNum, finalNum)` when consensus tells us there's been an update there.
What about TD information? is that from the old times or still useful?

Reply:
Why is the `td` on block end, rather than block start? 
does a non-nil `err` mean that the block is invalid? If so, is the `td` always nil?

Reply:
> Why is the td on block end, rather than block start?

JSON-RPC returns state at block end so it makes to have it on `OnBlockEnd` (it builds up as the execution happen) soI think it's why it was like that on our Firehose patch.

But yeah, works fine on `OnBlockStart` too.

> does a non-nil err mean that the block is invalid? If so, is the td always nil?

That's how I would see it, `err != nil` implies block was not committed and `td == nil`. Moving `td` to `OnBlockStart` (and with meaning that it's `td` at block start), we could make a simpler `OnBlockEnd(err)` indeed.

Reply:
Yeah that's good. I moved `td` to `OnBlockStart` and it means total difficulty prior to given block. I.e. you need to add with `block.Difficulty`.

Reply:
Also added finalized and safe headers to `OnBlockStart`.

> This however introduces some "flow" constraint if done using OnChainUpdate because we must sent the initial value before the first OnBlockStart

Re this tho it's no problem since on startup we load these markers from the database and could emit an event. The reason I suggested that approach was because finalized and safe markers stay the same for most blocks and change once every few minutes.